### DISCUSSION TITLE
Rebuilding static pages every day.

### DISCUSSION THREAD

**Discussion by titong0** [vercel/next.js#46555] [Help]
There's a route of artist/:id. I want to build the page once per day and at the other day build it again. 

1. SSR: the db calls the server does are a bit expensive and the response is always the same during a day. 

2. SSR with cache: I'm not entirely sure how this would work, I think the solution is somewhere in here but I don't know how to implement it correctly. Maybe adding a dynamic 'max-age' that equals the seconds left for the day to finish?

Then I tried using SSG with fallback: blocking

3. SSG with generatedAt prop: it's extremely fast but I'm not sure how to know wether a page is stale. Current implementation is returning a 'generatedAt' prop in getStaticProps and check in the client wether it is stale. In which case a revalidate api route would be called.

4. (Did not work) SSG with weird trick: I tried storing in an array every page that is NOT stale, then I'd check this on every request via middleware and revalidate if the page is not present. In 'getStaticProps' I would push to a GENERATED_TODAY array and then return the page. After that, a Cron job would run everyday at 00:00 and empty the array. This did not work  because the array was being defined on every request which meant it was always empty.

Anyway I'm not sure where the correct answer is but I'm pretty confident it has to do with middleware or cache. Thanks!

**Comment by ciruz** [comment]
Do you know about [``revalidate``](https://nextjs.org/docs/api-reference/data-fetching/get-static-props#revalidate)?

**Reply by titong0** [reply]
Yes! The main problem is I don't want to revalidate all pages at the same time. Only when it is necessary

**Reply by ciruz** [reply]
Okay I gotcha, check this example out [Using On-Demand Revalidation](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#using-on-demand-revalidation)

**Reply by titong0** [reply]
That is the solution that 3. Describes, however I'm asking for help regarding manually revalidating, I'm trying to revalidate on request but only when the page content is stale.

**Comment by ikapo** [comment]
@titong0 I too am looking for a solution to this, and found something that's supposed to work.
1. Create an API route that revalidates all pages you need (in my case only "/", but in your case you can query the DB to see which ones you need to validate) using `res.revalidate`
2. Create a cron job on vercel that calls that api route once every 24 hours

The only issue is that although I can confirm SSG occured, the response from the server is still stale.
If I call the API route from my local computer instead of the cron job revalidates and sends a non-stale response though, so I'm thinking maybe some cron-job headers are not invalidating the cache

UPDATE: This approach seems to work if the cron job calls an edge runtime API route, which in turn calls the normal API route.
A pretty annoying hack whatever works ig 🤷🏻

**Reply by titong0** [reply]
Hey! While this is very close to what I was intending, the main problem is I have thousands of pages, so I want to revalidate them when they are stale, but only when requested. What I want is a page to be stale for months as long as it is not requested.
